### DISCUSSION TITLE
How can I connect Oracle database use SID instead serviceName?

### DISCUSSION THREAD

**Discussion by tuhalang** [godror/godror#285] [Q&A] [answered]
For docs, the connection string should be: ${user}/${pass}@${host}:${port}/${serviceName}
But when I connect to a database using SID, I always got an error: ORA-12541: TNS:no listener.
I thought It related to a connection string because in Java if you connect using SID the connection string must be changed to ${user}/${pass}@${host}:${port}:${serviceName}

Anyone can help me?

**Comment by tgulacsi** [comment]
https://godror.github.io/godror/doc/connection.html may help.

**Reply by tuhalang** [reply]
I did follow it but still got an error

**Comment by tgulacsi** [comment]
"Note that database “system identifiers“ (SIDs) cannot be used in Easy Connect or Easy
Connect Plus syntax"

AFAIK those SIDs are only defined in tnsnames.ora and should be used directly: ${user}/${passw}@${sid}

**Reply by tuhalang** [reply]
where should the hostname be placed?

**Reply by tgulacsi** [reply]
Either SID, or hostname

**Reply by tuhalang** [reply]
you mean that: ${user}/${pass}@${host}:${port}/${sid} ?

**Reply by tgulacsi** [reply]
Nope. Your version is for serviceName. For SID, see https://github.com/godror/godror/discussions/285#discussioncomment-5884231

**Reply by tuhalang** [reply]
Not really understand your comment, Can you give me a connection string example?

**Reply by tgulacsi** [reply]
/@sid

**Reply by tuhalang** [reply]
@tgulacsi I tried it but still got an error. This is my connection string: **TEST_1/abc123@localhost:1521/@XE**

**Reply by Xe** [reply]
hi, can I help you?

**Reply by tuhalang** [reply]
@Xe sorry, in my connection string have a "@" character, so it mentions you.

**Reply by tgulacsi** [reply]
If the sid is Xe ( it is in tnsnames.ora) then use "TEST_1/abc123@xe".

**Reply by tuhalang** [reply]
@tgulacsi So how to driver know the hostname and port?

**Reply by tuhalang** [reply]
This is fully my connection which has an error: sqlx.Connect("godror", "TEST_1/abc123@localhost:1521/XE")

**Comment by tgulacsi** [comment]
It is not a godror problem. How do you connect with sqlplus?

**Reply by tuhalang** [reply]
![image](https://github.com/godror/godror/assets/33905113/2c4ebd57-3fbd-4ba3-a3fe-da0c44243313)
I connected successfully by IDEA

**Reply by cjbj** [reply]
SID were obsoleted a long time ago in Oracle.  A service name is preferred.   SIDs can't be used in the 'hostname:port/servicename' syntax. Recent versions of XE will have the service name of XEPDB1 or XECDB

**Reply by tuhalang** [reply]
@cjbj I set up for testing, so I use an oracle db in docker then I need connect using SID

**Reply by cjbj** [reply, marked as answer]
I don't think you do !  But if you insist, then don't use the easy connect syntax.  This may help: https://github.com/godror/godror/blob/main/doc/connection.md#-connection-strings

In case you didn't know, the latest version is 23c.  See https://www.oracle.com/database/technologies/free-downloads.html Also see the announcement https://www.oracle.com/news/announcement/oracle-database-23c-free-developer-release-2023-04-03/

**Reply by tgulacsi** [reply]
"TEST_1/abc123@(DESCRIPTION=(ADDRESS=(PROTOCOL=TCP)(HOST=localhost)(PORT=1521))(CONNECT_DATA=(SERVICE_NAME=xepdb1)))"
or 
"TEST_1/abc123@localhost:1521/xepdb1"
or
with xecdb
or 
with XE
as service name?

**Comment by AChangFeng** [comment]
In case of anyone still don't kown how to connect to Oracle database use SID, here is an example:
```go
func NewOraDB(host, port, svcName, sid, username, password string, timeout int) (*sql.DB, error) {
	dsn := godror.ConnectionParams{}
	connString := ""
	if svcName != "" {
		connString = fmt.Sprintf("%s:%s/%s?connect_timeout=%d", host, port, svcName, timeout)
	} else if sid != "" {
		connString = fmt.Sprintf(`(DESCRIPTION=
(ADDRESS=(PROTOCOL=TCP)(HOST=%s)(PORT=%s))
(CONNECT_DATA=(SID=%s))
(CONNECT_TIMEOUT=%ds)(RETRY_COUNT=3)(RETRY_DELAY=2)(TRANSPORT_CONNECT_TIMEOUT=%ds))`, host, port, sid, timeout, timeout)
	} else {
		return nil, errors.New("service name or sid must be set")
	}
	dsn.ConnectString = connString
	dsn.Username = username
	dsn.Password = godror.NewPassword(password)
	dsn.NoTZCheck = true

	dsnString := dsn.StringWithPassword()
	return sql.Open("godror", dsnString)
}
```

**Reply by cjbj** [reply]
The reason to prefer using Service Names instead of SIDS, is that Service Names allow various Oracle Database features to be enabled, such as Application Continuity.  So if you have a choice, avoid using SIDs.
